crates/services/p2p/src/gossipsub/config.rs

```diff
@@ -54,150 +102,104 @@ pub fn default_gossipsub_builder() -> GossipsubConfigBuilder {
-        mesh_failure_penalty_weight: -1.0,
-        mesh_failure_penalty_decay: 0.5,
-
-        ///  Invalid messages
-        ///  This is the number of invalid messages in the topic.
-        ///  The value of the parameter is the square of the counter, decaying with
-        ///  `invalid_message_deliveries_decay`.
-        ///  The weight of the parameter MUST be negative (or zero to disable).
-        invalid_message_deliveries_weight: -1.0,
-        invalid_message_deliveries_decay: 0.3,
-    }
+fn initialize_topic_score_params(
+    topic_weight: f64,
+    // The expected rate of message propagation.
+    expected_message_rate: f64,
+) -> TopicScoreParams {
+    // The decay time for the first message delivered score, set to 100 times the epoch duration.
+    // This means that the score given for first message deliveries will decay over this time period.
+    let first_message_decay_time = EPOCH * 100;
+
+    let mut params = TopicScoreParams::default();
+
+    params.topic_weight = topic_weight;
+
+    // The "quantum" of time spent in the mesh, set to the duration of a slot.
+    // This is the smallest unit of time for which we track a peer's presence in the mesh.
+    params.time_in_mesh_quantum = SLOT;
+
+    params.time_in_mesh_cap = 3600.0 / params.time_in_mesh_quantum.as_secs_f64();
+    params.time_in_mesh_weight = 10.0 / params.time_in_mesh_cap;
+
+    params.first_message_deliveries_decay =
+        score_parameter_decay(first_message_decay_time);
+
+    params.first_message_deliveries_cap = decay_convergence(
+        params.first_message_deliveries_decay,
+        2.0 * expected_message_rate / MESH_SIZE as f64,
+    );
+    params.first_message_deliveries_weight = 40.0 / params.first_message_deliveries_cap;
+
+    params.mesh_message_deliveries_weight = 0.5;
+    params.mesh_message_deliveries_threshold = 3.0;
+    params.mesh_message_deliveries_decay = 0.1;
+    params.mesh_message_deliveries_cap = MAX_POSITIVE_SCORE * 0.5;
+    params.mesh_message_deliveries_window = Duration::from_secs(60);
+    params.mesh_message_deliveries_activation = Duration::from_secs(120);
+    params.mesh_failure_penalty_decay = 0.1;
+    params.mesh_failure_penalty_weight = -0.05;
+
+    params.invalid_message_deliveries_weight = -MAX_POSITIVE_SCORE / params.topic_weight;
+    params.invalid_message_deliveries_decay = score_parameter_decay(EPOCH * 50);
+
+    params
 }
 
-/// This function takes in `max_gossipsub_score` and sets it to `topic_score_cap`
-/// The reasoning is because app-specific is set to 0 so the max peer score
-/// can be the score of the topic score cap
-fn initialize_peer_score_params() -> PeerScoreParams {
-    PeerScoreParams {
-        // topics are added later
-        topics: Default::default(),
+fn score_parameter_decay(decay_time: Duration) -> f64 {
+    let ticks = decay_time.as_secs_f64() / DECAY_INTERVAL.as_secs_f64();
+    DECAY_TO_ZERO.powf(1.0 / ticks)
```

Comment:
Are there any scenarios where `ticks == 0`?

Reply:
there shouldn't be, this is only a helper function used during setup, it will not called ever again.